src/generic_mtrie_impl.hpp

```diff
@@ -0,0 +1,461 @@
+            if (next.table[c]->is_redundant ()) {
+                LIBZMQ_DELETE (next.table[c]);
+
+                zmq_assert (live_nodes > 0);
+                --live_nodes;
+            } else {
+                //  The node is not redundant, so it's a candidate for being
+                //  the new min/max node.
+                //
+                //  We loop through the node array from left to right, so the
+                //  first non-null, non-redundant node encountered is the new
+                //  minimum index. Conversely, the last non-redundant, non-null
+                //  node encountered is the new maximum index.
+                if (c + min < new_min)
+                    new_min = c + min;
+                if (c + min > new_max)
+                    new_max = c + min;
+            }
+        }
+    }
+
+    zmq_assert (count > 1);
+
+    //  Free the node table if it's no longer used.
+    if (live_nodes == 0) {
+        free (next.table);
+        next.table = NULL;
+        count = 0;
+    }
+    //  Compact the node table if possible
+    else if (live_nodes == 1) {
+        //  If there's only one live node in the table we can
+        //  switch to using the more compact single-node
+        //  representation
+        zmq_assert (new_min == new_max);
+        zmq_assert (new_min >= min && new_min < min + count);
+        generic_mtrie_t *node = next.table[new_min - min];
+        zmq_assert (node);
+        free (next.table);
+        next.node = node;
+        count = 1;
+        min = new_min;
+    } else if (new_min > min || new_max < min + count - 1) {
+        zmq_assert (new_max - new_min + 1 > 1);
+
+        generic_mtrie_t **old_table = next.table;
+        zmq_assert (new_min > min || new_max < min + count - 1);
+        zmq_assert (new_min >= min);
+        zmq_assert (new_max <= min + count - 1);
+        zmq_assert (new_max - new_min + 1 < count);
+
+        count = new_max - new_min + 1;
+        next.table =
+          (generic_mtrie_t **) malloc (sizeof (generic_mtrie_t *) * count);
+        alloc_assert (next.table);
+
+        memmove (next.table, old_table + (new_min - min),
+                 sizeof (generic_mtrie_t *) * count);
+        free (old_table);
+
+        min = new_min;
+    }
+}
+
+template <typename T>
+bool zmq::generic_mtrie_t<T>::rm (prefix_t prefix_,
+                                  size_t size_,
+                                  value_t *pipe_)
+{
+    return rm_helper (prefix_, size_, pipe_);
+}
+
+template <typename T>
+bool zmq::generic_mtrie_t<T>::rm_helper (prefix_t prefix_,
+                                         size_t size_,
+                                         value_t *pipe_)
+{
+    if (!size_) {
+        // TODO pipes can only be NULL here, if we are at the top level, i.e. 
+        // rm was already called with size_ == 0. This could be checked in rm, 
+        // and here we could just have an assertion or naught
```

Comment:
If I understand correctly, this could also happen for duplicate unsubs - as above, not specified as illegal in the API docs and, I think, harmless as well

Reply:
turns out duplicated unsubs for the same pipe can never happen, as XSUB filters them at the origin (it keeps a parallel trie)

Reply:
No, it could not happen even for duplicate unsubscriptions, since the tree is always reorganized after a removal (I have some doubts that this behaviour is optimal, but this is a separate issue).